### ISSUE TITLE
Pre-Processing User Queries with LLM for Improved Similarity Search

### ISSUE THREAD

**Issue by khalilxg** [ragapp/ragapp#43]
Add a feature to LlamaIndex where user queries are first processed by an LLM to better match the indexed data. This processed query is then sent to the vector database for improved search results.

Workflow:

User submits a query.
Query is processed by an LLM to align with indexed data.
Processed query is sent to the vector database.
Search results are returned to the user.
Benefits:

Enhanced search accuracy.
More relevant results.
Improved user experience.
Example:

Original Query: "How to fix login issues?"
LLM Processed Query: "Resolving authentication errors in user access"
Thank you for considering this request.

**Comment by marcusschiesser** [comment]
@khalilxg good idea, we could add that as an option. Do you have a snippet using your code?

**Comment by khalilxg** [comment]
Not yet honestly still looking for other workarounds

**Comment by Rauljauregi** [comment]
I have read this article :

https://www.cmswire.com/customer-experience/improving-gen-ai-accuracy-with-retrieval-augmented-generation/

It provides a great way to imporve RAG results. One of the steps is what you suggest.

Thank you,

**Comment by marcusschiesser** [comment]
seems like there is already a paper on doing first a plan for the retrieval, called PlanRAG: https://arxiv.org/abs/2406.12430

**Comment by Rauljauregi** [comment]
Anyway based on the testings I am doing I believe actually, there is some kind of query  optimization that is done. Maybe it can be improved. Congrats

**Comment by marcusschiesser** [comment]
@Rauljauregi note if you're not using any tools, RAGapp is actually using the `Condense Plus Context Mode` (see https://docs.llamaindex.ai/en/stable/examples/chat_engine/chat_engine_condense_plus_context/) which is doing a LLM preprocessing for the question (condenses the conversation and latest user message to a standalone question).

What we're not doing yet is adding hybrid search (coming soon)
